_posts/2024-03-20-strimzi-kraft-migration.md

```diff
@@ -0,0 +1,265 @@
+my-cluster   3                        3                     True    KRaftDualWriting   
+my-cluster   3                        3                     True    KRaftPostMigration   
+my-cluster   3                        3                     True    KRaftPostMigration   True
+```
+
+In this status, the cluster is still working in "dual-write" mode and the user can validate that everything is working fine before deciding to finalize the migration or rolling back to use ZooKeeper.
+If the `Kafka` custom resource has the `inter.broker.protocol.version` and `log.message.format.version` parameters set in the `spec.kafka.config` section, and because they are not supported in KRaft, the operator reports warnings into the status (see `WARNINGS` column).
+The user has to remove the parameters at the end of the process causing an additional rolling of the nodes.
+
+Once the metadata state is in `KRaftPostMigration` state, you have two options how you can continue:
+
+* [finalize the migration](#migration-finalization) to have a full KRaft cluster.
+* [rollback the process](#rollback-to-zookeeper) to keep using ZooKeeper.
+
+#### Migration finalization
+
+At this point, the user can finalize the migration by changing the `strimzi.io/kraft` annotation to `enabled`.
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: Kafka
+metadata:
+  name: my-cluster
+  annotations:
+    strimzi.io/kraft: enabled # updated from "migration" to "enabled" to finalize the migration process
+    strimzi.io/node-pools: enabled
+# ...
+```
+
+You can do that by annotating the custom resource.
+
+```shell
+kubectl annotate kafka my-cluster strimzi.io/kraft="enabled" --overwrite
+```
+
+The operator will configure the KRaft controllers without the ZooKeeper connection details, with the migration flag disabled and roll them.
+When both brokers and controllers are working in KRaft mode, the operator will delete all the resources related to ZooKeeper as well.
+At this point, the user can delete the `.spec.zookeeper` section from the `Kafka` custom resource because it is not needed anymore.
+
+During the finalization, the cluster metadata state moves to the final `KRaft`.
+
+```shell
+NAME         DESIRED KAFKA REPLICAS   DESIRED ZK REPLICAS   READY   METADATA STATE   WARNINGS
+...
+my-cluster   3                        3                     True    KRaftPostMigration   True
+my-cluster   3                        3                     True    PreKRaft             True
+my-cluster   3                        3                     True    KRaft                True
+my-cluster                                                  True    KRaft                True
+my-cluster                                                  True    KRaft                
+```
+
+As you can see, just a couple of changes on the `strimzi.io/kraft` annotation make the migration really straight forward.
+
+#### Rollback to ZooKeeper
```

Comment:
```suggestion
#### Rolling back the migration to ZooKeeper
```

Reply:
What about `Rolling back the cluster to ZooKeeper`?

Reply:
Yeah, I prefer Mickael suggestion.